### ISSUE TITLE
'Could not load file or assembly 'System.Fabric.Management, Version=6.0.0.0' when running .NET Core 2.0 app

### ISSUE THREAD

**Issue by mkosieradzki** [Azure/service-fabric-issues#638]
When running .NET Core 2.0 Stateful Service which uses FabricClient
```
System.IO.FileNotFoundException: 'Could not load file or assembly 'System.Fabric.Management, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.'
```
is thrown.

I am using:
```
<PackageReference Include="Microsoft.ServiceFabric" Version="6.1.232-preview" />
<PackageReference Include="Microsoft.ServiceFabric.Services" Version="3.0.232-preview" />
```

And in SCD .NET Core 2.0 application there is no System.Fabric.Management.dll.
Might be related to #563

**Comment by ashishnegi** [comment]
@mkosieradzki Please share your code project or minimum issue reproducing project where you are facing this. This will help in faster resolution.

**Comment by mkosieradzki** [comment]
@ashishnegi I think this one will be quite easy to repro. I will try to create one soon.

**Comment by ashishnegi** [comment]
@mkosieradzki  Thanks. Also please share your SF runtime version, SDK version, Env like Linux or windows machine.

**Comment by mkosieradzki** [comment]
@ashishnegi 
OK this is repro (runtime 6.0.232, DevCluster, Windows, SDK 3.0.232-preview):

1. Create Stateful Service in .NET Core 2.0
2. In StatefulService subclass add code below
3. Create folder C:\test123 (empty)
4. Run service on dev cluster - you will get the aforementioned exception

```
        private static readonly XNamespace F = "http://schemas.microsoft.com/2011/01/fabric";

        protected async override Task RunAsync(CancellationToken cancellationToken)
        {
            var client = new FabricClient();
            var clusterManifest = XElement.Parse(await client.ClusterManager.GetClusterManifestAsync());
            var imageStoreConnectionString = (string)clusterManifest.Element(F + "FabricSettings").Elements(F + "Section").Single(x => (string)x.Attribute("Name") == "Management").Elements(F + "Parameter").Single(x => (string)x.Attribute("Name") == "ImageStoreConnectionString").Attribute("Value");
            client.ApplicationManager.CopyApplicationPackage(imageStoreConnectionString, @"C:\test123", "test123");
        }
```

**Comment by raunakpandya** [comment]
#640

**Comment by karataliu** [comment]
I got same error as described when calling 'CopyApplicationPackage', is there a fix?

Runtime: .NET Core 2.0 project using nuget Microsoft.ServiceFabric 6.1.232-preview on windows

**Comment by SamuelMcAravey-NVP** [comment]
+1 here. Windows with the just released SDK 3.0.456, Visual Studio 15.5.6, and trying to deploy to the local cluster. I have attached my project that reproduces the error.

I created a brand new stateless Asp.Net Core 2.0 project on .NET Core, hit F5 and got this error:
```
[AspNetCoreRepro.zip](https://github.com/Azure/service-fabric-issues/files/1675059/AspNetCoreRepro.zip)

Copying application to image store...
Copy-ServiceFabricApplicationPackage : Could not load type 
'System.Fabric.Management.ImageBuilder.FabricImageBuilderReservedDirectoryException' from assembly 
'System.Fabric.Management, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
At C:\Program Files\Microsoft SDKs\Service 
Fabric\Tools\PSModule\ServiceFabricSDK\Publish-NewServiceFabricApplication.ps1:244 char:9
+         Copy-ServiceFabricApplicationPackage -ApplicationPackagePath  ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Copy-ServiceFabricApplicationPackage], TypeLoadException
    + FullyQualifiedErrorId : CopyApplicationPackageErrorId,Microsoft.ServiceFabric.Powershell.CopyApplicationPackage
 
Finished executing script 'Deploy-FabricApplication.ps1'.
Time elapsed: 00:00:06.5139778
The PowerShell script failed to execute.```

**Comment by Tadimsky** [comment]
I'm getting the same error with the latest SDK...

**Comment by raunakpandya** [comment]
@SamuelMcAravey-NVP - The error you seeing is not .NET Core specific, seems like a different issue. If so, can you please open a different issue for tracking? 
@Tadimsky  - What error you getting? The CopyApplicationPackage/RemoveApplicationPackage/CopyClusterPackage/RemoveClusterPackage API are currently not supported on .NET Core SDK (v3.0). They would be made available in the upcoming release.

**Comment by Tadimsky** [comment]
@raunakpandya This was hitting F5 in Visual Studio in an SF project that contains two .Net Core 2.0 services.

**Comment by deanmccrae** [comment]
I also got this error (below), and have managed to get past it.

This error occurred I think, after the Visual Studio upgrade to 15.5.6. My own project failed with this error (I needed to first upgrade the project formats) during deploying to the local cluster (simply F5). Also tried to simply create a brand new service fabric application with stateless service - none of my own code - same thing. Happened also on my colleague's machine - however a third person in the team was not affected by it.

Also, I have Powershell scripts that use the Service Fabric low-level CmdLets (not the scripts that Visual Studio uses for deployment) and these were failing also.

I have managed to get it working now, with these steps:

1. Uninstall the SF SDK (and runtime) using Windows Programs and Features.
2. Uninstall Visual Studio altogether (a repair did not help).
3. Re-install Visual Studio - includes 'Service Fabric Tools' under the Cloud components.
4. Re-install the Service Fabric SDK 3.0.456 using Web Platform Installer for VS2017.
5. Restart.

The problem has gone away: Visual Studio can deploy and so can my Powershell scripts.

I did not find a more light-weight way to resolve it.

Also, I am uncertain what the 'Service Fabric Tools' in Visual Studio installation actually gives you that you don't get already from the Service Fabric SDK in the Web Platform Installer - because the latter seems to install the Powershell modules used by Visual Studio deployments, and also the general Service Fabric CmdLets. Perhaps someone here can shed light on this.

--------------------------

Copy-ServiceFabricApplicationPackage : Could not load type 
'System.Fabric.Management.ImageBuilder.FabricImageBuilderReservedDirectoryException' from assembly 
'System.Fabric.Management, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
At C:\Program Files\Microsoft SDKs\Service 
Fabric\Tools\PSModule\ServiceFabricSDK\Publish-NewServiceFabricApplication.ps1:244 char:9
+         Copy-ServiceFabricApplicationPackage -ApplicationPackagePath  ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Copy-ServiceFabricApplicationPackage], TypeLoadException
    + FullyQualifiedErrorId : CopyApplicationPackageErrorId,Microsoft.ServiceFabric.Powershell.CopyApplicationPackage

**Comment by inech** [comment]
@deanmccrae Unfortunately, reinstalling didn't help me, so I have rolled back to the previous version of SF SDK.

 I have uninstalled (from control panel):

- Microsoft Azure Service Fabric 6.1.456
- Microsoft Azure Service Fabric SDK 3.0.456

And then installed 'Microsoft Azure Service Fabric SDK 2.8.232' from Web Platform Installer (runtime installation is bundled inside).

As a result, now I can deploy SF applications while keeping the latest version of Visual Studio (15.5.6).

**Comment by ThiemeNL** [comment]
Any update on this? I have exactly the same issue with netcore 2.0.

**Comment by PepijnGramberg** [comment]
In the release notes of 6.2/3.1 it is stated that this issue is fixed. But after upgrading everything to the new version: (6.2/3.1).274.9494, runtime/sdk and package references, this problem still persists. We want to manage applications from a .netcoreapp client, but all actions with the FabricClient.ApplicationManager still return this exception.

``` System.IO.FileNotFoundException: Could not load file or assembly 'System.Fabric.Management, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. 
The system cannot find the file specified.File name: 'System.Fabric.Management, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'   
at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMarkHandle stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName, ObjectHandleOnStack type, ObjectHandleOnStack keepalive)   
at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean loadTypeFromPartialName)   
at System.RuntimeType.GetType(String typeName, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark& stackMark)   
at System.Type.GetType(String typeName, Boolean throwOnError)   
at System.Fabric.ImageStoreProxy..ctor(String imageStoreConnectionString, String[] hostEndpoints, SecurityCredentials credential)   
at System.Fabric.FabricClient.GetImageStore(String imageStoreConnectionString)   
at System.Fabric.FabricClient.ApplicationManagementClient.RemoveApplicationPackage(String imageStoreConnectionString, String applicationPackagePathInImageStore) 
```

It is simple to repro using the following code in a netcoreapp2.0:
```
new FabricClient().ApplicationManager.RemoveApplicationPackage("file:C:\\SfDevCluster\\Data\\ImageStoreShare", "appType");
```

Did the current "fix" in the release notes relate to an alternate bug? or is it simply not fixed yet?

We have been waiting for a fix for this for 6 months already #640 and it has become a real bottleneck for us, when can we expect to get this fixed?

**Comment by ashishnegi** [comment]
@PepijnGramberg The fix is not yet released. It is scheduled for early next week. I will update the thread here once fix is available publicly.
Thanks for your patience. Due to other issues, we could not release it sooner.

**Comment by ashishnegi** [comment]
@PepijnGramberg @ThiemeNL @inech @mkosieradzki @karataliu The fix is present in latest nupkg : 6.2.283. This change took too long to come out. We appreciate your patience. Let me know how it works for you.

**Comment by ThiemeNL** [comment]
@ashishnegi it is too long ago, I worked around the issue. I cannot reproduce it anymore. Sorry.

**Comment by ashishnegi** [comment]
Closing the issue as fix is in 6.2.283.
